src/briefcase/integrations/android_sdk.py

```diff
@@ -212,61 +248,305 @@ def verify_emulator(self):
+            # The first line is header information.
+            # Each subsequent line is a single device descriptor.
+            devices = {}
+            for line in output.split("\n")[1:]:
+                parts = re.sub(r"\s+", " ", line).split(" ")
+
+                details = {}
+                for part in parts[2:]:
+                    key, value = part.split(":")
+                    details[key] = value
+
+                if parts[1] == "device":
+                    name = details["device"]
+                    authorized = True
+                else:
+                    name = "Unknown device (not authorized for development)"
+                    authorized = False
+
+                devices[parts[0]] = {
+                    "name": name,
+                    "authorized": authorized,
+                }
+
+            return devices
+        except subprocess.CalledProcessError:
+            raise BriefcaseCommandError("Unable to obtain Android device list")
 
-The device ID is the value in the first column of output - it will be either:
+    def select_target_device(self, device_or_avd):
+        """
+        Select a device to be the target for actions
 
-  * a ~12-16 character alphanumeric string (for a physical device); or
-  * a value like `emulator-5554` (for an emulator).
+        Interrogates the system to get the list of available devices
 
-If you do not see any devices, you can create and start an emulator by running:
+        If the user has specified a device at the command line, that device
+        will be validated, and then automatically selected
+
+        :param device_or_avd: The device or AVD to target. Can be a physical
+            device id (a hex string), an emulator id ("emulator-5554"), or
+            an emulator AVD name ("@robotfriend"). If ``None``, the user will
+            be asked to select a device from the list available.
+        :returns: A tuple containing ``(device, name, avd)``. ``avd``
+            will only be provided if an emulator with that AVD is not currently
+            running. If ``device`` is null, a new emulator should be created.
+        """
+        # Get the list of attached devices (includes running emulators)
+        running_devices = self.devices()
+
+        # Choices is an ordered list of options that can be shown to the user.
+        # Each device should appear only once, and be keyed by AVD only if
+        # a device ID isn't available.
+        choices = []
+        # Device choices is the full lookup list. Devices can be looked up
+        # by any valid key - ID *or* AVD.
+        device_choices = {}
+
+        # Iterate over all the running devices.
+        # If the device has an AVD, use ADB to get the emulator AVD name.
+        # If it is a physical device, use the device name.
+        # Keep a log of all running AVDs
+        running_avds = {}
+        for d, details in sorted(running_devices.items(), key=lambda d: d[1]["name"]):
```

Comment:
Pedantic: Why sorted here?

Reply:
Python3.5 test ordering. Python 3.5 doesn't guarantee reliable dictionary order, so sorting is the easiest way to make it's always processed in the same order.